sefaria/model/linker/ref_resolver.py

```diff
@@ -435,28 +625,56 @@ def _get_unrefined_ref_part_matches_for_title_context(self, context_ref: Optiona
+        while curr_node is not None:
+            path_nodes.append(curr_node)
+            curr_node = getattr(curr_node, 'parent', None)
+        mutation_set = ContextMutationSet()
+        for path_node in reversed(path_nodes):
+            raw_mutations = getattr(path_node, 'ref_resolver_context_mutations', None)
+            if not raw_mutations:
+                continue
+            parsed_mutations = self._parse_context_mutation_data(raw_mutations, path_node)
+            if parsed_mutations:
+                mutation_set.add_mutations(parsed_mutations)
+        return mutation_set if len(mutation_set) > 0 else None
 
-        Modifies `raw_ref` with updated ref_parts
-        """
-        swapped_ref_parts = []
-        term_matcher = self.get_term_matcher()
-        if context_swap_map is None: return
-        for part in raw_ref.raw_ref_parts:
-            # TODO assumes only one match in term_matches
-            term_matches = term_matcher.match_term(part)
-            found_match = False
-            for match in term_matches:
-                if match.slug not in context_swap_map: continue
-                swapped_ref_parts += [TermContext(schema.NonUniqueTerm.init(slug)) for slug in context_swap_map[match.slug]]
-                found_match = True
-                break
-            if not found_match: swapped_ref_parts += [part]
-        raw_ref.parts_to_match = swapped_ref_parts
+    @staticmethod
+    def _parse_context_mutation_data(raw_mutations: Iterable[dict], node: schema.SchemaNode) -> List[ContextMutation]:
+        parsed: List[ContextMutation] = []
+        node_title = node.full_title() if hasattr(node, "full_title") else getattr(node, "key", str(node))
+        for raw_mutation in raw_mutations:
+            if not isinstance(raw_mutation, dict):
+                logger.warning("ref_resolver.context_mutation.invalid_format", node=node_title, data=raw_mutation)
+                continue
+            op_token = raw_mutation.get("op")
+            input_terms = raw_mutation.get("input_terms", [])
+            output_terms = raw_mutation.get("output_terms", [])
+            try:
+                op = ContextMutationOp(op_token)
+            except Exception:
+                logger.warning("ref_resolver.context_mutation.invalid_op", node=node_title, op=op_token)
+                continue
+            if not isinstance(input_terms, (list, tuple)) or not all(isinstance(term, str) for term in input_terms):
+                logger.warning("ref_resolver.context_mutation.invalid_input_terms", node=node_title, input_terms=input_terms)
+                continue
+            if not isinstance(output_terms, (list, tuple)) or not all(isinstance(term, str) for term in output_terms):
+                logger.warning("ref_resolver.context_mutation.invalid_output_terms", node=node_title, output_terms=output_terms)
+                continue
+            try:
+                parsed.append(ContextMutation(op, input_terms, output_terms))
```

Comment:
we can assume the ContextMutation is valid since we'll validate it on SchemaNode save